File: problems/leetcode/LargestNumber.java
class Solution {
    public String largestNumber(int[] nums) {
        Queue<String> q = new PriorityQueue<>(nums.length, (a, b) -> (b + a).compareTo(a + b));

        for (int n : nums) q.add(Integer.toString(n));

        if (q.peek().equals("0")) {
            return "0";
        }

        StringBuilder largest = new StringBuilder();
        while (!q.isEmpty()) largest.append(q.poll());

        return largest.toString();
    }
}

File: problems/leetcode/largest-number.cc
// https://leetcode.com/problems/largest-number/
/*
Brute Force:
============
1. Convert all to string.
2. Inverse sort array based on digit order.
3. return "".join(nums)

TC: O(N log N)
SC: O(N), where N = len(nums).
*/

class Solution {
public:
    string largestNumber(vector<int>& nums) {
        vector<std::string> stringNums;
        for (const auto num : nums) {
            stringNums.push_back(std::to_string(num));
        }

        std::sort(stringNums.begin(), stringNums.end(),
            [](const std::string& a, const std::string& b) {
                return a + b > b + a; });

        if (stringNums[0] == "0") {
            return "0";
        }

        std::string largest;
        for (const auto& num : stringNums) {
            largest += num;
        }

        return largest;
    }
};
